libvast/src/arrow_table_slice.cpp

```diff
@@ -585,387 +603,196 @@ class index_applier {
-        }
-      } else if constexpr (std::is_same_v<Type, address_type>) {
-        if constexpr (std::is_same_v<Array, address_array>) {
-          co_yield address_at(
-            static_cast<const arrow::FixedSizeBinaryArray&>(*arr.storage()), i);
-        } else {
-          type_mismatch();
-        }
-      } else if constexpr (std::is_same_v<Type, subnet_type>) {
-        if constexpr (std::is_same_v<Array, subnet_array>) {
-          co_yield subnet_at(
-            static_cast<const arrow::StructArray&>(*arr.storage()), i);
-        } else {
-          type_mismatch();
-        }
-      } else if constexpr (std::is_same_v<Type, string_type>) {
-        if constexpr (std::is_same_v<Array, arrow::StringArray>) {
-          co_yield string_at(arr, i);
-        } else {
-          type_mismatch();
-        }
-      } else if constexpr (std::is_same_v<Type, pattern_type>) {
-        if constexpr (std::is_same_v<Array, pattern_array>) {
-          co_yield pattern_view{string_at(
-            static_cast<const arrow::StringArray&>(*arr.storage()), i)};
-        } else {
-          type_mismatch();
-        }
-      } else if constexpr (std::is_same_v<Type, list_type>) {
-        if constexpr (std::is_same_v<Array, arrow::ListArray>) {
-          co_yield list_at(type.value_type(), arr, i);
-        } else {
-          type_mismatch();
-        }
-      } else if constexpr (std::is_same_v<Type, map_type>) {
-        if constexpr (std::is_same_v<Array, arrow::MapArray>) {
-          co_yield map_at(type.key_type(), type.value_type(), arr, i);
-        } else {
-          type_mismatch();
-        }
-      } else if constexpr (std::is_same_v<Type, record_type>) {
-        if constexpr (std::is_same_v<Array, arrow::StructArray>) {
-          // TODO: record_at is expensive, breaking the columnar processing
-          co_yield record_at(type, arr, i);
-        } else {
-          type_mismatch();
-        }
-      } else {
-        static_assert(detail::always_false_v<Type>, "unhandled vast type");
-      }
-    }
+    else
+      co_yield value_at(type, arr, i);
   }
 }
 
-auto values(const type& type, const arrow::Array& array)
+auto values(const type& type, const std::same_as<arrow::Array> auto& array)
   -> detail::generator<data_view> {
-  auto f = [](const concrete_type auto& type,
-              const auto& array) -> detail::generator<data_view> {
-    for (auto&& result : values(type, array)) {
-      if (result)
-        co_yield *result;
+  const auto f = []<concrete_type Type>(
+                   const Type& type,
+                   const arrow::Array& array) -> detail::generator<data_view> {
+    for (auto&& result :
+         values(type, caf::get<type_to_arrow_array_t<Type>>(array))) {
+      if (!result)
+        co_yield {};
       else
-        co_yield caf::none;
+        co_yield std::move(*result);
     }
   };
-  return caf::visit(f, type, array);
+  return caf::visit(f, type, detail::passthrough(array));
```

Comment:
Note to the reviewer: `array` cannot be captured in the lambda, as the generator coroutine starts suspended and it is thus undefined behavior to access it. The new `detail::passthrough` helper function lifts a const-lvalue reference into a wrapper type that goes through `caf::visit` untouched.